### PR TITLE
feat: Earner Status Manager

### PR DESCRIPTION
**Pull request by deluca-mike**

### CODE REVIEW

File: test/utils/WrappedMTokenHarness.sol [hunk 1]
```diff
@@ -8,8 +8,9 @@ contract WrappedMTokenHarness is WrappedMToken {
     constructor(
         address mToken_,
         address migrationAdmin_,
-        address claimRecipientManager
-    ) WrappedMToken(mToken_, migrationAdmin_, claimRecipientManager) {}
+        address earnerStatusManagerManager_,
+        address claimRecipientManager_
+    ) WrappedMToken(mToken_, migrationAdmin_, earnerStatusManagerManager_, claimRecipientManager_) {}
```
Comment (PierrickGT) [inline test/utils/WrappedMTokenHarness.sol:13]:
```suggestion
        address earnerStatusManager_,
        address claimRecipientManager_
    ) WrappedMToken(mToken_, migrationAdmin_, earnerStatusManager_, claimRecipientManager_) {}
```

**Review by PierrickGT — APPROVED** [review]

File: src/EarnerStatusManager.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    /// @inheritdoc IEarnerStatusManager
+    address public immutable registrar;
+
+    /// @dev Mapping of account to earner status.
+    mapping(address account => bool earnerStatus) internal _earnerStatuses;
+
+    /* ============ Modifiers ============ */
+
+    modifier onlyEarnerStatusAdmin() {
+        _revertIfNotEarnerStatusAdmin();
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply src/EarnerStatusManager.sol:33]: Whilke correct, anything in a modifier gets copied into where the modifier is used. Keeping `_revertIfNotEarnerStatusAdmin` as it's own function means `setEarnerStatus` and `setEarnerStatuses` will only have those jumps inlined, and not the whole check and revert code.

**Review by toninorair — COMMENTED** [review]
pending product approval

File: src/WrappedMToken.sol [hunk 3]
```diff
@@ -680,9 +681,7 @@ contract WrappedMToken is IWrappedMToken, Migratable, ERC20Extended {
      * @return isApproved_ True if the account_ is a Registrar-approved earner, false otherwise.
      */
     function _isApprovedEarner(address account_) internal view returns (bool isApproved_) {
-        return
-            IRegistrarLike(registrar).get(_EARNERS_LIST_IGNORED) != bytes32(0) ||
-            IRegistrarLike(registrar).listContains(_EARNERS_LIST, account_);
+        return IEarnerStatusManager(earnerStatusManager).earnerStatusFor(account_);
```
Comment (toninorair) [inline src/WrappedMToken.sol:684]: measure gas savings for inlining EarnerStatusManager logic within WrappedMToken vs calling external contract

File: src/WrappedMToken.sol [hunk 4]
```diff
@@ -595,6 +579,60 @@ contract WrappedMToken is IWrappedMToken, Migratable, ERC20Extended {
...
+        accountInfo_.isEarning = true;
+        accountInfo_.lastIndex = currentIndex_;
+
+        uint240 balance_ = accountInfo_.balance;
+
+        _addTotalEarningSupply(balance_, currentIndex_);
+
+        unchecked {
+            totalNonEarningSupply -= balance_;
+        }
```
Comment (PierrickGT) [inline src/WrappedMToken.sol:603]: Could be nice to add a comment explaining why it can't underflow.

File: src/WrappedMToken.sol [hunk 5]
```diff
@@ -595,6 +579,60 @@ contract WrappedMToken is IWrappedMToken, Migratable, ERC20Extended {
...
+        accountInfo_.isEarning = false;
+        accountInfo_.lastIndex = 0;
+
+        uint240 balance_ = accountInfo_.balance;
+
+        _subtractTotalEarningSupply(balance_, currentIndex_);
+
+        unchecked {
+            totalNonEarningSupply += balance_;
+        }
```
Comment (PierrickGT) [inline src/WrappedMToken.sol:631]: Could be nice to add a comment explaining why it can't overflow.

**Review by PierrickGT — APPROVED** [review]

File: .github/workflows/test-integration.yml [hunk 6]
```diff
@@ -20,12 +20,6 @@ jobs:
       - name: Install Foundry
         uses: foundry-rs/foundry-toolchain@v1
 
-      - name: Run Forge build
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply .github/workflows/test-integration.yml:23]: Running tests already builds. What we need is a dedicated check for contract sizes with production options.

File: src/WrappedMToken.sol (hunk 4 above)
Comment (deluca-mike) [reply src/WrappedMToken.sol:603]: Perrhaps, but we have 4 of these unchecked decrements in the code, and they are unrelated to this PR (this code was just moved/internalized, not changed).

**Pull request closed without merging**